server/src/main/java/io/deephaven/server/barrage/BarrageStreamGenerator.java

```diff
@@ -251,6 +263,87 @@ public void forEachStream(Consumer<InputStream> visitor) throws IOException {
         public boolean isViewport() {
             return this.viewport != null;
         }
+
+        public final StreamReader.StreamReaderOptions options() {
+            return new StreamReader.StreamReaderOptions(this.options);
+        }
+
+        public final RowSet keyspaceViewport() {
+            return this.keyspaceViewport;
+        };
+
+    }
+
+    /**
+     * Obtain a View of this StreamGenerator that can be sent to a single snapshot requestor.
+     *
+     * @param options serialization options for this specific view
+     * @param viewport is the position-space viewport
+     * @param keyspaceViewport is the key-space viewport
+     * @param subscribedColumns are the columns subscribed for this view
+     * @return a MessageView filtered by the snapshot properties that can be sent to that subscriber
+     */
+    @Override
+    public SnapshotView getSnapshotView(final BarrageSnapshotOptions options,
+            @Nullable final RowSet viewport,
+            @Nullable final RowSet keyspaceViewport,
+            @Nullable final BitSet subscribedColumns) {
+        return new SnapshotView(this, options, viewport, keyspaceViewport, subscribedColumns);
+    }
+
+    /**
+     * Obtain a Full-Snapshot View of this StreamGenerator that can be sent to a single snapshot requestor.
+     *
+     * @param options serialization options for this specific view
+     * @return a MessageView filtered by the snapshot properties that can be sent to that subscriber
+     */
+    @Override
+    public SnapshotView getSnapshotView(BarrageSnapshotOptions options) {
+        return getSnapshotView(options, null, null, null);
+    }
+
+    public static class SnapshotView implements View {
+        public final BarrageStreamGenerator generator;
+        public final BarrageSnapshotOptions options;
+        public final RowSet viewport;
+        public final RowSet keyspaceViewport;
+        public final BitSet subscribedColumns;
+        public final boolean hasAddBatch;
+        public final boolean hasModBatch;
+
+        public SnapshotView(final BarrageStreamGenerator generator,
+                final BarrageSnapshotOptions options,
+                @Nullable final RowSet viewport,
+                @Nullable final RowSet keyspaceViewport,
+                @Nullable final BitSet subscribedColumns) {
+            this.generator = generator;
+            this.options = options;
+            this.viewport = viewport;
+            this.keyspaceViewport = keyspaceViewport;
+            this.subscribedColumns = subscribedColumns;
+            this.hasModBatch = false;
+            this.hasAddBatch = generator.rowsIncluded.original.isNonempty();
+        }
+
+        @Override
+        public void forEachStream(Consumer<InputStream> visitor) throws IOException {
+            ByteBuffer metadata = generator.getSnapshotMetadata(this);
```

Comment:
In our discussion the other day, I think we discussed getting rid of metadata on snapshots altogether. If we don't, we should at least narrow the snapshot metadata down to a meaningful subset of existing fields.